internal/containerd/sidecar.go

```diff
@@ -0,0 +1,232 @@
+		return fmt.Errorf("failed to get hostname: %v", err)
+	}
+
+	filter := fmt.Sprintf("labels.io.kubernetes.container.name==%s,labels.io.kubernetes.pod.name==%s", hostname, hostname)
+
+	for {
+		conts, err := s.client.Containers(ctx, filter)
+		if err != nil {
+			return fmt.Errorf("failed to get containers: %v", err)
+		}
+
+		if len(conts) == 1 {
+			s.primaryContainerID = conts[0].ID()
+			return nil
+		}
+		if len(conts) > 1 {
+			return fmt.Errorf("too many containers matching filters: %d", len(conts))
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(time.Second * 1):
+		}
+	}
+}
+
+func (s *containerdSidecar) Commit(ctx context.Context, image string) (string, error) {
+	imageRef, err := refDocker.ParseDockerRef(image)
+	if err != nil {
+		return "", err
+	}
+	digest, err := commit.Commit(ctx, s.client, s.primaryContainerID, &commit.Opts{
+		Ref: imageRef.String(),
+	})
+	if err != nil {
+		return "", err
+	}
+	return digest.String(), nil
+}
+
+func (s *containerdSidecar) Upload(ctx context.Context, fileName string) error {
+	container, err := s.client.LoadContainer(ctx, s.primaryContainerID)
+	if err != nil {
+		return err
+	}
+	info, err := container.Info(ctx)
+	if err != nil {
+		return err
+	}
+	mounts, err := s.client.SnapshotService(info.Snapshotter).Mounts(ctx, s.primaryContainerID)
+	if err != nil {
+		return err
+	}
+	srcFile, err := os.Open(fileName)
+	if err != nil {
+		return err
+	}
+	defer srcFile.Close()
+	return mount.WithTempMount(ctx, mounts, func(root string) error {
+		dstFileName := filepath.Join(root, srcFile.Name())
+		var dstFile *os.File
+		dstFile, err = os.Create(dstFileName)
+		if err != nil {
+			return err
+		}
+		_, err = io.Copy(dstFile, srcFile)
+		if err != nil {
+			return err
+		}
+		return dstFile.Sync()
+	})
+}
+
+func (s *containerdSidecar) BuildImage(ctx context.Context, fileName, image string) error {
+	// TODO(cezarsa): build must work, requires running buildkit daemon
+	return errors.New("build not supported yet")
+}
+
+func (s *containerdSidecar) TagAndPush(ctx context.Context, baseImage string, destinationImages []string, reg sidecar.RegistryConfig, w io.Writer) error {
+	baseRef, err := refDocker.ParseDockerRef(baseImage)
+	if err != nil {
+		return err
+	}
+	image, err := s.client.ImageService().Get(ctx, baseRef.String())
+	if err != nil {
+		return err
+	}
+
+	authorizer := remoteDocker.NewDockerAuthorizer(
+		remoteDocker.WithAuthClient(pushHTTPClient),
+		remoteDocker.WithAuthCreds(func(string) (string, string, error) {
+			return reg.RegistryAuthUser, reg.RegistryAuthPass, nil
+		}))
+	registryHosts := remoteDocker.ConfigureDefaultRegistries(
+		remoteDocker.WithAuthorizer(authorizer),
+		remoteDocker.WithClient(pushHTTPClient),
+		remoteDocker.WithPlainHTTP(func(host string) (ret bool, err error) {
+			local, err := remoteDocker.MatchLocalhost(host)
+			if local {
+				return local, err
+			}
+			return strings.HasPrefix(host, "192.168."), nil
```

Comment:
@cezarsa could we have better form to detect http mode ?, people could use 10/8 network in their homes.

Reply:
This was a hack I used to allow a registry on my local network, I think I'll remove this logic completely, leaving only `remoteDocker.MatchLocalhost`.